### PR TITLE
Dimension compression UNCOMPRESSED should not use CompressedVSizeIntsIndexedWriter

### PR DESCRIPTION
**Pull request by navis**
Generally Indices for dimension are not compressed well and makes bigger (~50%) sized output. But by setting dimension compressed UNCOMPRESSED, the output is always pad with 64K, which is default size of CompressedPools. I think we should use VSizeIndexedIntsWriter instead of CompressedVSizeIntsIndexedWriter in UNCOMPRESSED configuration.

--
 
We are testing druid in industry realm and I've unexpectedly found that the size reduction of data could be important appeal point of using druid (1/10 ~ 1/30 for parquet/snappy but druid easily achieves 1/100). Trying to make this more apparent I've found in the most cases compressing index makes bigger sized footprint than original size of it. But disabling compression made 64K padding attached and makes it bigger again. This is try for fixing that.

### CODE REVIEW

**Comment by xvrl** [comment]
@navis will this be backwards compatible with segments that were written using CompressedVSizeIntsIndexedWriter? Druid needs to maintain backwards compatibility with any segments that were written by a stable release in the past.

File: processing/src/main/java/io/druid/segment/IndexMergerV9.java [hunk 1]
```diff
@@ -335,7 +335,8 @@ private void makeDimensionColumns(
 
     long startTime = System.currentTimeMillis();
     final BitmapSerdeFactory bitmapSerdeFactory = indexSpec.getBitmapSerdeFactory();
-    final CompressedObjectStrategy.CompressionStrategy compressionStrategy = indexSpec.getDimensionCompressionStrategy();
+    final boolean compressed =
```
Comment (xvrl) [inline processing/src/main/java/io/druid/segment/IndexMergerV9.java]: any reason we are only changing `IndexMergerV9`?

Comment (navis) [inline processing/src/main/java/io/druid/segment/IndexMergerV9.java]: @xvrl It's related to what you've mentioned above (keeping changes just in new V9 merger). It writes `version` value in header properly and backward compatible.

Comment (navis) [inline processing/src/main/java/io/druid/segment/IndexMergerV9.java]: @xvrl Should I apply this also to IndexMerger?

**Comment by fjy** [comment]
@xvrl  any more comments?

**Comment by KurtYoung** [comment]
Why the old version will use CompressedVSizeIntsIndexedWriter in UNCOMPRESSED configuration?
If you configured "uncompressed", the IndexSpec will return null when we call getDimensionCompressionStrategy().
And if dimCompression is null, IndexMergerV9 will return either VSizeIndexedWriter or VSizeIndexedIntsWriter depends on whether is multi-value dimension

**Comment by navis** [comment]
@KurtYoung `getDimensionCompressionStrategy` does not return null.

**Comment by KurtYoung** [comment]
image
@navis

**Comment by navis** [comment]
@KurtYoung 
````
@Test
public void testUncompressed() throws Exception
{
  final IndexSpec spec = new IndexSpec(null, "uncompressed", null);
  Assert.assertEquals(CompressedObjectStrategy.CompressionStrategy.UNCOMPRESSED, spec.getDimensionCompressionStrategy());
}
```

`"uncompressed"` returns `CompressionStrategy.UNCOMPRESSED`, making `compressed=true`

**Comment by KurtYoung** [comment]
@navis I mean the original version of code base will return null when we call getDimensionCompressionStrategy(). You changed the logic of getDimensionCompressionStrategy() so it will return CompressionStrategy.UNCOMPRESSED now.
My point is IMO the original behavior is right, i'm curious why you saw the "uncompressed" config not working when you first firing this PR.

**Comment by navis** [comment]
@KurtYoung Strange. I was being nut or worked on dirty branch. Thanks.

**Comment by KurtYoung** [comment]
@navis never mind, i think you made this logic more clear and readable. Depending on "null" is not a really good choice.

**Pull request closed without merging**